File: 1582-design-browser-history/1582-design-browser-history.cpp
class BrowserHistory {
public:
vector<string>v;
int i=0;
    BrowserHistory(string homepage) {
        v.push_back(homepage);
        i=0;
    }
    
    void visit(string url) {
        v.erase(v.begin() + i + 1, v.end());
        v.push_back(url);
        i = v.size() - 1;
    }
    
    string back(int steps) {
          i = max(0, i - steps);
        return v[i];
    }
    
    string forward(int steps) {
         i = min((int)v.size() - 1, i + steps);
        return v[i];
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1582-design-browser-history/1582-design-browser-history.py
class Node:
    def __init__(self, url='', prev=None, next=None):
        self.url = url
        self.prev = prev
        self.next = next

class BrowserHistory:
    def __init__(self, homepage: str):
        self.curNode = Node(url=homepage)

    def visit(self, url: str) -> None:
        newNode = Node(url=url, prev=self.curNode)
        self.curNode.next = newNode
        self.curNode = newNode

    def back(self, steps: int) -> str:
        while self.curNode.prev and steps > 0:
            self.curNode = self.curNode.prev
            steps -= 1
        return self.curNode.url

    def forward(self, steps: int) -> str:
        while self.curNode.next and steps > 0:
            self.curNode = self.curNode.next
            steps -= 1
        return self.curNode.url
